### ISSUE TITLE
create function return  enum as array  core dump  when orca is open

### ISSUE THREAD

**Issue by shenfowang** [greenplum-db/gpdb#11880]
### Greenplum version or build
PostgreSQL 9.4.24 (Greenplum Database 6.16.0 build commit:5650be2b79197fed564dca8d734d10f2a76b876c Open Source) on x86_64-unknown-li
### OS version and uname -a

CentOS Linux release 7.3.1611 (Core)
Linux node129 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

### autoconf options used ( config.status --config )
null

### Installation information ( pg_config )
BINDIR = /home/gpadmin/greenplum-db-6.16.0/bin
DOCDIR = /home/gpadmin/greenplum-db-6.16.0/share/doc/postgresql
HTMLDIR = /home/gpadmin/greenplum-db-6.16.0/share/doc/postgresql
INCLUDEDIR = /home/gpadmin/greenplum-db-6.16.0/include
PKGINCLUDEDIR = /home/gpadmin/greenplum-db-6.16.0/include/postgresql
INCLUDEDIR-SERVER = /home/gpadmin/greenplum-db-6.16.0/include/postgresql/server
LIBDIR = /home/gpadmin/greenplum-db-6.16.0/lib
PKGLIBDIR = /home/gpadmin/greenplum-db-6.16.0/lib/postgresql
LOCALEDIR = /home/gpadmin/greenplum-db-6.16.0/share/locale
MANDIR = /home/gpadmin/greenplum-db-6.16.0/man
SHAREDIR = /home/gpadmin/greenplum-db-6.16.0/share/postgresql
SYSCONFDIR = /home/gpadmin/greenplum-db-6.16.0/etc/postgresql
PGXS = /home/gpadmin/greenplum-db-6.16.0/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--disable-debug-extensions' '--disable-tap-tests' '--enable-orca' '--with-zstd' '--with-gssapi' '--with-libxml' '--with-perl' '--with-python' '--with-openssl' '--with-pam' '--with-ldap' '--with-extra-version= Open Source' '--prefix=/usr/local/greenplum-db-6.16.0' '--mandir=/usr/local/greenplum-db-6.16.0/man' 'CC=gcc' 'CFLAGS=-O3 -fargument-noalias-global -fno-omit-frame-pointer -g'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/home/gpadmin/greenplum-db-6.16.0/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -O3 -fargument-noalias-global -fno-omit-frame-pointer -g -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/home/gpadmin/greenplum-db-6.16.0/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/usr/local/greenplum-db-6.16.0/lib',--enable-new-dtags -L/home/gpadmin/greenplum-db-6.16.0/lib
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lpam -lrt -lgssapi_krb5 -lzstd -lrt -lcrypt -ldl -lm -L/home/gpadmin/greenplum-db-6.16.0/lib
VERSION = PostgreSQL 9.4.24

### Expected behavior
CREATE TYPE rainbow AS ENUM('red','orange','yellow','green','blue','indigo','violet');
CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
RETURNS TABLE (ae anyenum, aa anyarray) AS $$
SELECT $1, array[$2, $3]
$$ LANGUAGE SQL STABLE;
SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);

execute all sql should ok 

### Actual behavior
coredump

### Step to reproduce the behavior

```
postgres=# select version();
                                                                                                      version

-------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------
 PostgreSQL 9.4.24 (Greenplum Database 6.16.0 build commit:5650be2b79197fed564dca8d734d10f2a76b876c Open Source) on x86_64-unknown-li
nux-gnu, compiled by gcc (GCC) 6.4.0, 64-bit compiled on Apr 22 2021 21:21:23
(1 row)

postgres=# CREATE TYPE rainbow AS ENUM('red','orange','yellow','green','blue','indigo','violet');
CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
RETURNS TABLE (ae anyenum, aa anyarray) AS $$
SELECT $1, array[$2, $3]
$$ LANGUAGE SQL STABLE;
SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);
CREATE TYPE
postgres=# CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
postgres-# RETURNS TABLE (ae anyenum, aa anyarray) AS $$
postgres$# SELECT $1, array[$2, $3]
postgres$# $$ LANGUAGE SQL STABLE;
CREATE FUNCTION
postgres=# SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!> \q
```

note: if set optimizer to off , execute sql is ok

coredump stack
```
(gdb) bt
#0  0x00007f2f051ca23b in raise () from /lib64/libpthread.so.0
#1  0x0000000000bf4bc0 in StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=<optimized out>, postgres_signal_arg=11)
    at elog.c:5548
#2  <signal handler called>
#3  operator new (line=398, filename=0x1271fc8 "CTranslatorUtils.cpp", mp=0x5, size=48)
    at ../../../../src/backend/gporca/libgpos/include/gpos/memory/CMemoryPool.h:280
#4  gpdxl::CTranslatorUtils::ResolvePolymorphicTypes (mp=mp@entry=0x35e4ec8, mdid_array=mdid_array@entry=0x37f7cc8,
    arg_types_list=<optimized out>, funcexpr=funcexpr@entry=0x37cf880) at CTranslatorUtils.cpp:398
#5  0x0000000000fc5fd6 in gpdxl::CTranslatorUtils::ConvertToCDXLLogicalTVF (mp=0x35e4ec8, md_accessor=0x7ffd0d43a280,
    id_generator=0x37b47f8, rte=rte@entry=0x37cf690) at CTranslatorUtils.cpp:317
#6  0x0000000000fe6807 in gpdxl::CTranslatorQueryToDXL::TranslateTVFToDXL (this=0x37b46a8, rte=0x37cf690, rt_index=1)
    at CTranslatorQueryToDXL.cpp:3576
#7  0x0000000000fdfc66 in gpdxl::CTranslatorQueryToDXL::TranslateFromExprToDXL (this=this@entry=0x37b46a8,
    from_expr=from_expr@entry=0x37cfb90) at CTranslatorQueryToDXL.cpp:2922
#8  0x0000000000fe10ea in TranslateSelectProjectJoinToDXL (group_clause=0x0, output_attno_to_colid_mapping=0x37b8b38,
    sort_group_attno_to_colid_mapping=0x37b8138, from_expr=0x37cfb90, target_list=0x37cfc28, this=0x37b46a8)
    at CTranslatorQueryToDXL.cpp:608
#9  gpdxl::CTranslatorQueryToDXL::TranslateGroupingSets (this=this@entry=0x37b46a8, from_expr=0x37cfb90, target_list=0x37cfc28,
    group_clause=0x0, has_aggs=<optimized out>, sort_grpref_to_colid_mapping=sort_grpref_to_colid_mapping@entry=0x37b8138,
    output_attno_to_colid_mapping=0x37b8638) at CTranslatorQueryToDXL.cpp:2218
#10 0x0000000000fe152d in gpdxl::CTranslatorQueryToDXL::TranslateSelectQueryToDXL (this=this@entry=0x37b46a8)
    at CTranslatorQueryToDXL.cpp:557
#11 0x0000000000fe2cb6 in gpdxl::CTranslatorQueryToDXL::TranslateQueryToDXL (this=0x37b46a8) at CTranslatorQueryToDXL.cpp:659
#12 0x0000000000ff7864 in COptTasks::OptimizeTask (ptr=0x7ffd0d43ad90) at COptTasks.cpp:559
#13 0x0000000000ce8466 in gpos::CTask::Execute (this=this@entry=0x3798fe8) at CTask.cpp:130
#14 0x0000000000ce93f8 in gpos::CWorker::Execute (this=0x7ffd0d43ac40, task=task@entry=0x3798fe8) at CWorker.cpp:81
#15 0x0000000000ce792d in gpos::CAutoTaskProxy::Execute (this=this@entry=0x7ffd0d43ac70, task=task@entry=0x3798fe8)
    at CAutoTaskProxy.cpp:286
#16 0x0000000000ce9ff1 in gpos_exec (params=params@entry=0x7ffd0d43ad00) at _api.cpp:246
#17 0x0000000000ff5a11 in COptTasks::Execute (func=func@entry=0xff7630 <COptTasks::OptimizeTask(void*)>,
    func_arg=func_arg@entry=0x7ffd0d43ad90) at COptTasks.cpp:247
#18 0x0000000000ff6c3e in COptTasks::GPOPTOptimizedPlan (query=query@entry=0x36e0330,
    gpopt_context=gpopt_context@entry=0x7ffd0d43ad90) at COptTasks.cpp:774
---Type <return> to continue, or q <return> to quit---
#19 0x0000000000ff8f01 in GPOPTOptimizedPlan (had_unexpected_failure=0x7ffd0d43ae0f, query=0x36e0330) at CGPOptimizer.cpp:58
#20 GPOPTOptimizedPlan (query=query@entry=0x36e0330, had_unexpected_failure=had_unexpected_failure@entry=0x7ffd0d43ae0f)
    at CGPOptimizer.cpp:214
#21 0x00000000009d0d93 in optimize_query (parse=parse@entry=0x35c05b8, boundParams=boundParams@entry=0x0) at orca.c:138
#22 0x00000000009c241d in standard_planner (parse=0x35c05b8, cursorOptions=0, boundParams=0x0) at planner.c:238
#23 0x00000000009c2775 in planner (parse=parse@entry=0x35c05b8, cursorOptions=cursorOptions@entry=0,
    boundParams=boundParams@entry=0x0) at planner.c:200
#24 0x0000000000a885b1 in pg_plan_query (boundParams=0x0, cursorOptions=0, querytree=0x35c05b8) at postgres.c:959
#25 pg_plan_queries (cursorOptions=0, boundParams=0x0, querytrees=<optimized out>) at postgres.c:1018
#26 exec_simple_query (
    query_string=0x35bf038 "SELECT * FROM return_enum_as_array('red'::rainbow, 'green'::rainbow, 'blue'::rainbow);")
    at postgres.c:1748
#27 0x0000000000a8af4f in PostgresMain (argc=<optimized out>, argv=argv@entry=0x359ea30, dbname=<optimized out>,
    username=<optimized out>) at postgres.c:5242
#28 0x00000000006b1a05 in BackendRun (port=0x35c9e40) at postmaster.c:4811
#29 BackendStartup (port=0x35c9e40) at postmaster.c:4468
#30 ServerLoop () at postmaster.c:1948
#31 0x0000000000a10a12 in PostmasterMain (argc=argc@entry=6, argv=argv@entry=0x359ca20) at postmaster.c:1518
#32 0x00000000006b5791 in main (argc=6, argv=0x359ca20) at main.c:245

```

**Comment by d** [comment]
@shenfowang  I cannot reproduce this on master, have you tried?

**Comment by shenfowang** [comment]
@d  beacuse memory overflow, maybe can or cannot reproduce in some cases in different branch.

for example: 
1. using  https://github.com/greenplum-db/gpdb/releases/download/6.16.0/open-source-greenplum-db-6.16.0-rhel7-x86_64.rpm  on centos 7.3 always cause coredump
2.  compile source code (6x_stable branch ) with CFLAGS='-O0 -g3' , can not  reproduce this problem on centos7.3
3. compile source code (6x_stable branch ) with CFLAGS='-O2 -g' , can reproduce this problem on centos7.3

then I find problem cause，look my pull requests：
Fix array overflow for function argument types in orca #12051

----
    Fix array overflow for function argument types in orca

    in function CTranslatorUtils::ResolvePolymorphicTypes(), the size
    of array arg_types is num_args, and it is smaller than total_args,
    this causes a memory corruption and a backend crash. Fixed by
    correcting the array size "total_args"(num_args + num_return_args).

    Fixes GitHub issue #11880

**Comment by chrishajas** [comment]
Merged fix to 6X (and master) in https://github.com/greenplum-db/gpdb/commit/2f45387e41c5a01cd18905e57c43662730b97e6c, thanks for the report and contribution!
